syntax = "proto3";
package gen;

import "Address.proto";

enum RequestMessageType {
    CREATE_REQUEST = 0;
    UPDATE_REQUEST = 1;
    READ_REQUEST = 2;
    DELETE_REQUEST = 3;
}

message CreateRequestFields {
    string value = 1;
}

message UpdateRequestFields {
    string value = 1;
}

message ReadRequestFields {
}

message DeleteRequestFields {
}

message RequestMessage {
    RequestMessageType messageType = 1;
    Address sourceAddress = 2;
    Address destinationAddress = 3;
    string id = 4;
    string key = 5;

    oneof customFields {
        CreateRequestFields createFields = 6;
        UpdateRequestFields updateFields = 7;
        ReadRequestFields readFields = 8;
        DeleteRequestFields deleteFields = 9;
    }
}

enum ResponseMessageType {
    CREATE_RESPONSE = 0;
    UPDATE_RESPONSE = 1;
    READ_RESPONSE = 2;
    DELETE_RESPONSE = 3;
}

message CreateResponseFields {
}

message UpdateResponseFields {
}

message ReadResponseFields {
}

message DeleteResponseFields {
}


message ResponseMessage {
    ResponseMessageType messageType = 1;
    Address sourceAddress = 2;
    Address destinationAddress = 3;
    string id = 4;
    string originalMessageId = 5;
    int32 responseCode = 6;

    oneof customFields {
        CreateResponseFields createFields = 7;
        UpdateResponseFields updateFields = 8;
        ReadResponseFields readFields = 9;
        DeleteResponseFields deleteFields = 10;
    }
}